### PR TITLE
Add Feature: Printing Absolute Path in Context

### PR DESCRIPTION
**Pull request by HelinXu**
`ic.configureOutput(includeContext=True, absPath=True)`

`absPath`, False by default, can be used to control whether `ic()`'s output includes the absolute path of the file where `ic()` is called, provided `includeContext == True`. This is useful when `ic()` is called from multiple files in a project, with files potentially having same names but different paths. Moreover, for most editors such as VSCode, a clickable link to the line where `ic()` is called is provided. E.g. a click on `/absolute/path/to/example.py:12` brings you to the line of the debugging call. It is really handy when you have debugging outputs from multiple files and want to easily jump to the line where the debugging call is made.

### CODE REVIEW

File: icecream/icecream.py [hunk 1]
```diff
@@ -330,9 +332,13 @@ def _getContext(self, callFrame, callNode):
...
         parentFunction = frameInfo.function
-        filename = basename(frameInfo.filename)
 
-        return filename, lineNumber, parentFunction
+        if self.absPath:
+            filepath = realpath(frameInfo.filename)
+            return filepath, lineNumber, parentFunction
+        else:
+            filename = basename(frameInfo.filename)
+            return filename, lineNumber, parentFunction
```
Comment (alexmojaki) [inline icecream/icecream.py:341]:
```suggestion
        filepath = (realpath if self.absPath else basename)(frameInfo.filename)
        return filepath, lineNumber, parentFunction
```

Comment (HelinXu) [reply icecream/icecream.py:341]: That makes total sense😃
I'm updating this first thing tomorrow

**Comment by gruns** [comment]
@HelinXu please rename the `absPath` parameter to `contextAbsPath` so it's clear what path, exactly, is being controlled via the absolute path parameter. beyond that, lgtm 🙌 

@alexmojaki zooming out, i think replacing the `includeContext` param, `prefix` param, and soon-to-be `contextAbsPath` param, of `configureOutput()` with one, single function that lets users customize all the content before the ic output, ie replaces both the prefix and context, could improve output flexibility. that way users can fully customize ic's output to their exact liking, albeit in a manner that exposes ic's internals like the `callFrame` and `callNode` currently used in `IceCreamDebugger._formatContext()`

we could then provide a pre-built function that replicates the current behavior. something like

```python
from icecream import ic, createPrefixFn

_formatPrefix = createPrefixFn(prefix='ic |', includeContext=True, contextAbsPath=True)
ic.prefixFn  = _formatPrefix
```

conversely we could also just improve the documentation so people feel its easier to subclass `IceCreamDebugger()` and extend `_formatContext()` to their own liking and wait for more people to express a desire to customize the prefix and/or context

thoughts?

**Comment by HelinXu** [comment]
Done. Thanks for the advice 🙌

**Comment by gruns** [comment]
awesome. thank you, @HelinXu!

last step is to add appropriate tests for `contextAbsPath` in https://github.com/gruns/icecream/blob/master/tests/test_icecream.py

see https://github.com/gruns/icecream/blob/master/tests/test_icecream.py#L440-L468 for how `includeContext` is tested, along with https://github.com/gruns/icecream/blob/master/tests/test_icecream.py#L108-L140 which tests lines for the inclusion of context

with tests, ill merge your amazing work here! 🙌

**Comment by HelinXu** [comment]
Sure, I'm working on that.

One more thing, there're actually two slightly different handlings when it comes to absolute path, i.e. `abspath()` and `realpath()`. They are basically dealing with symbolic links differently:

```python
def abspath(path):
    """Return an absolute path."""
    path = os.fspath(path)
    if not isabs(path):
        if isinstance(path, bytes):
            cwd = os.getcwdb()
        else:
            cwd = os.getcwd()
        path = join(cwd, path)
    return normpath(path)


# Return a canonical path (i.e. the absolute location of a file on the
# filesystem).

def realpath(filename):
    """Return the canonical path of the specified filename, eliminating any
symbolic links encountered in the path."""
    filename = os.fspath(filename)
    path, ok = _joinrealpath(filename[:0], filename, {})
    return abspath(path)
```

I'm using `realpath()` for now🤔. Any suggestions on this?

**Comment by HelinXu** [comment]
tests added :)

```
test_icecream.py::TestIceCream::testContextAbsPathMultiLine PASSED                                 [ 19%]
# Output with absolute path can easily exceed line width, so no assert line num here.
test_icecream.py::TestIceCream::testContextAbsPathSingleLine PASSED                                [ 22%]
```

I've written `lineIsAbsPathContext(line)`, similar to `lineIsContext(line)`. It seems "with/without abs path output" doubles test cases for several config options, but I did not bother to create tests all of them as it's too redundant.

File: tests/test_icecream.py [hunk 2]
```diff
@@ -467,6 +492,18 @@ def testIncludeContextMultiLine(self):
...
         assert pair == ('multilineStr', ic.argToStringFunction(multilineStr))
 
+    def testContextAbsPathMultiLine(self):
+        multilineStr = 'line1\nline2'
+        with configureIcecreamOutput(includeContext=True, contextAbsPath=True):
+            with disableColoring(), captureStandardStreams() as (out, err):
+                ic(multilineStr)
+
+        firstLine = err.getvalue().splitlines()[0]
+        assert lineIsAbsPathContext(firstLine) # TODO
```
Comment (gruns) [inline tests/test_icecream.py:502]: quick question: what remains TODO here? or is this an old comment and no longer applicable? 🙂

**Review by HelinXu — COMMENTED** [review]
I've resolved the unattended TODO and have modified MYFILEPATH according to the latest commit. Sorry for the carelessness XD

**Comment by gruns** [comment]
merged! 🎉 

thank you for your wonderful suggestion and, most importantly, your wonderful PR, @HelinXu 🙌

**Comment by HelinXu** [comment]
Hi, I'm curious if 2.1.3 is to be released soon, as I was kind of hoping to use this new feature :)

在2022-06-11 11:55:29，Ansgar ***@***.***写道：

merged! 🎉

thank you for your wonderful suggestion and, most importantly, your wonderful PR, @HelinXu🙌

**Pull request merged**